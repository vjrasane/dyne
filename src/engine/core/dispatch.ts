import { types as T, utils as U } from "dyne-commons";
import Effect from "../effects/effect";

export type Dispatch = (msg: any) => void;

export type Queues = {
  msg: any[];
  effect: Effect<any>[];
};

const dispatcher = <A>(queue: A[], queueable: A): void => {
  U.exists(queueable) && queue.push(queueable);
};

export const msgEventListener = (
  listener: T.Generator<object>,
  dispatch: Dispatch
): T.Sink => (...args: any[]): void => dispatch(listener(...args));

export const getDispatcher = () => {
  const queues: Queues = {
    msg: [],
    effect: []
  };

  const dispatch: Dispatch = (queueable: Effect<any> | object): void =>
    queueable instanceof Effect
      ? dispatcher(queues.effect, queueable)
      : dispatcher(queues.msg, queueable);

  return {
    queues,
    dispatch
  };
};
