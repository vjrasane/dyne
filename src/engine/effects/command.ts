import { utils as U } from "dyne-commons";
import { CmdFx, ProcessorF, PromiseableF, Processor, Failure } from "./types";
import Effect from "./effect";
import { Effectable } from "./return";
import { Dispatch } from "../core/dispatch";

export class Command<I, O> extends Effect<CmdFx<I>> {
  success: Processor<I, O>;
  failure: Failure<O>;

  constructor(cmd: CmdFx<I>, success?: Processor<I, O>, failure?: Failure<O>) {
    super(cmd);
    this.success = U.exists(success) ? success : U.identity;
    this.failure = U.exists(failure) ? failure : U.identity;
  }

  execute = (dispatch: Dispatch): Promise<void> =>
    processCommand(dispatch, this);
}

const processResult = (dispatch: Dispatch) => <I, O>(
  processor: Processor<I, O>,
  result: I
): void =>
  dispatch(
    U.isFunction(processor) ? (<ProcessorF<I, O>>processor)(result) : processor
  );

const processCommand = async <I, O>(
  dispatch: Dispatch,
  cmd: Command<I, O>
): Promise<void> => {
  const process: <I, O>(
    processor: Processor<I, O>,
    result: I
  ) => void = processResult(dispatch);

  try {
    const result = U.isFunction(cmd.effect)
      ? (<PromiseableF<I>>cmd.effect)()
      : cmd.effect;
    process(cmd.success, await result);
  } catch (error) {
    process(cmd.failure, error);
  }
};

export function Cmd<I, O>(
  cmd: CmdFx<I>,
  success?: Processor<I, O>,
  failure?: Failure<O>
): Command<I, O> {
  return new Command<I, O>(cmd, success, failure);
}

Cmd.wrap = <A>(wrappable: Effectable<A>): Effect<CmdFx<A>> => {
  if (wrappable instanceof Effect) {
    return wrappable;
  } else {
    return Cmd(wrappable);
  }
};
