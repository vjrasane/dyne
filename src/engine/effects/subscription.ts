import Effect from "./effect";
import { types as T } from "dyne-commons";
import { Cmd, Command } from "./command";
import { Dispatch, msgEventListener } from "../core/dispatch";
import { Subscriber, Canceller, Subscribe, Cancel } from "./types";

export class Subscription extends Effect<Subscriber> {
  canceller: Canceller;

  constructor(subscriber: Subscriber, canceller: Canceller) {
    super(subscriber);
    this.canceller = canceller;
  }

  execute = (dispatch: Dispatch): void => this.effect(dispatch);
  cancel = (): Command<void, void> => Cmd(this.canceller);
}

export const Sub = <A>(subscribe: Subscribe<A>, cancel: Cancel<A>) => {
  let sub: A;
  const subscriber: Subscriber = (dispatch: Dispatch): void => {
    sub = subscribe(dispatch);
  };
  const canceller: Canceller = (): void => cancel(sub);
  return new Subscription(subscriber, canceller);
};

Sub.interval = <A>(effect: Effect<A> | A, duration: number): Subscription =>
  Sub(
    (dispatch: Dispatch): NodeJS.Timer =>
      setInterval((): void => dispatch(effect), duration),
    (interval: NodeJS.Timer): void => clearInterval(interval)
  );

Sub.listen = (trigger: string, effect: T.Generator<object>): Subscription =>
  Sub(
    (dispatch: Dispatch): T.Sink => {
      const listener: T.Sink = msgEventListener(effect, dispatch);
      window.addEventListener(trigger, listener);
      return listener;
    },
    (listener: T.Sink): void => window.removeEventListener(trigger, listener)
  );
