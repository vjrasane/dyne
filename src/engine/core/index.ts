import { types as T } from "dyne-commons";
import { Setup, setup, Opts } from "./setup";
import { getDispatcher } from "./dispatch";

const UPDATE_INTERVAL = 1; /* milliseconds */
const VIEW_INTERVAL = 10; /* milliseconds */
const RENDER_INTERVAL = 10; /* milliseconds */
const EFFECT_INTERVAL = 1; /* milliseconds */

const intervals = {
  update: UPDATE_INTERVAL,
  view: VIEW_INTERVAL,
  render: RENDER_INTERVAL,
  effect: EFFECT_INTERVAL
};

export default <F, M>(setupObj: Setup<F, M>) => {
  const { dispatch, queues } = getDispatcher();
  const engine = setup(dispatch, queues, setupObj);
  return (opts: Opts<F>): void => {
    const procedures = engine(opts);
    Object.entries(procedures).forEach(([name, procedure]) =>
      setInterval(procedure, intervals[name])
    );
  };
};
