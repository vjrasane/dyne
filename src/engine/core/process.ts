import { utils as U } from "dyne-commons";
import Return from "../effects/return";
import { Updated, Update } from "./update";
import { Dispatch } from "./dispatch";

export default <A>(
  dispatch: Dispatch,
  updated: Updated<A>,
  previous?: A
): A => {
  const processed: Return<A> = Update.wrap(updated);
  processed.effects.forEach(e => dispatch(e));
  return U.exists(processed.model) ? processed.model : previous;
};
