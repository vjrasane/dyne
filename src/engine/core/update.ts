import { utils as U } from "dyne-commons";
import Return, { Effectable } from "../effects/return";
import Effect from "../effects/effect";
import { CmdFx } from "../effects/types";
import process from "./process";
import { Dispatch } from "./dispatch";

const throwUpdateError = (update: any): (() => never) =>
  U.throwError(
    new Error(
      `Attempted to pass message to 'update', but it was not a function, was: ${typeof update}`
    )
  );

export type Updated<A> = Return<A> | Effect<any> | CmdFx<A> | A;

export type Updater<A> = (msg: object, model: A) => A;

export type Update<A> = (msg: object, model: A) => Updated<A>;

export const Update = <A>(model: A, ...effects: Effectable<any>[]): Return<A> =>
  new Return(model, ...effects);

Update.wrap = <A>(wrappable: Updated<A>): Return<A> => {
  if (wrappable instanceof Return) {
    return wrappable;
  } else if (
    U.isFunction(wrappable) ||
    wrappable instanceof Promise ||
    wrappable instanceof Effect
  ) {
    return Update(null, wrappable);
  } else {
    return Update(<A>wrappable);
  }
};

export const getUpdater = <M>(
  dispatch: Dispatch,
  update: Update<M>
): Updater<M> => {
  if (U.isFunction(update)) {
    return (msg: object, model: M) =>
      process(dispatch, update(msg, model), model);
  }
  return throwUpdateError(update);
};
