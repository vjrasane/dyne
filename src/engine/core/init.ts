import { utils as U } from "dyne-commons";
import process from "./process";
import { Updated } from "./update";
import { Dispatch } from "./dispatch";

export type Init<F, M> = ((flags: F) => Updated<M>) | Updated<M>;

export type Initializer<F, M> = (flags: F) => M;

export const getInitializer = <F, M>(
  dispatch: Dispatch,
  init?: Init<F, M>
): Initializer<F, M> => {
  const initial: (flags: F) => Updated<M> = U.isFunction(init)
    ? <Initializer<F, M>>init
    : (_: F) => <Updated<M>>init;
  return (flags: F): M => process(dispatch, initial(flags));
};
