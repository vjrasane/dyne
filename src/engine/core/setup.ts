import { types as T, utils as U } from "dyne-commons";
import getRenderer, { Renderer } from "../render";
import { getInitializer, Initializer, Init } from "./init";
import { getUpdater, Updater, Update } from "./update";
import { getViewer, View } from "./view";
import { Dispatch, Queues } from "./dispatch";

export type Setup<F, M> = {
  init?: Init<F, M>;
  update?: Update<M>;
  view: View<M>;
};

export type Opts<F> = {
  node: HTMLElement;
  flags?: F;
};

export const setup = <F, M>(
  dispatch: Dispatch,
  queues: Queues,
  { init, update, view }: Setup<F, M>
) => {
  const initializer: Initializer<F, M> = getInitializer(dispatch, init);
  const updater: Updater<M> = getUpdater(dispatch, update);
  const viewer: View<M> = getViewer(view);
  const renderer: Renderer = getRenderer(dispatch);

  return ({ node, flags }: Opts<F>) => {
    let model: M = initializer(flags);

    const updateProcedure = () => {
      if (queues.msg.length > 0) {
        model = updater(queues.msg.shift(), model);
      }
    };

    let viewedModel: M;
    let virtualDom: T.VirtualDom;
    const viewVirtualDom = () => {
      virtualDom = viewer(model) || [];
      viewedModel = model;
    };

    const viewProcedure = () => {
      if (!virtualDom || (U.exists(model) && model !== viewedModel)) {
        viewVirtualDom();
      }
    };

    let renderedVirtualDom;
    const renderProcedure = () => {
      if (U.exists(virtualDom) && virtualDom !== renderedVirtualDom) {
        renderer(node, virtualDom, renderedVirtualDom);
        renderedVirtualDom = virtualDom;
      }
    };

    const effectProcedure = () => {
      if (queues.effect.length > 0) {
        const effect = queues.effect.shift();
        // execute effect with the message dispatcher
        return effect.execute(dispatch);
      }
    };

    return {
      update: updateProcedure,
      view: viewProcedure,
      render: renderProcedure,
      effect: effectProcedure
    };
  };
};
