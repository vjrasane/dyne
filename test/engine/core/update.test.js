import { getUpdater, Update } from "../../../src/engine/core/update";
import { Cmd } from "../../../src/engine/effects/command";
import { Sub } from "../../../src/engine/effects/subscription";

describe("update", () => {
  it("does not wrap an update", () => {
    const update = Update("model");
    expect(Update.wrap(update)).toBe(update);
  });

  it("wraps a command", () => {
    expect(Update.wrap(Cmd("effect"))).toMatchSnapshot();
  });

  it("wraps a subscription", () => {
    expect(Update.wrap(Sub.interval("effect", 100))).toMatchSnapshot();
  });

  it("wraps a model", () => {
    expect(Update.wrap("model")).toMatchSnapshot();
  });

  it("handles basic update", () => {
    const update = getUpdater(jest.fn(), (msg, model) => ({
      msg,
      model
    }));
    expect(update("msg", "model")).toMatchSnapshot();
  });

  it("dispatches command", () => {
    const dispatch = jest.fn();
    const cmd = Cmd("effect");
    const update = getUpdater(dispatch, () => cmd);
    expect(update("msg", "model")).toBe("model");
    expect(dispatch).toHaveBeenCalledWith(cmd);
  });

  it("dispatches subscription", () => {
    const dispatch = jest.fn();
    const sub = Sub.interval("effect", 100);
    const update = getUpdater(dispatch, () => sub);
    expect(update("msg", "model")).toBe("model");
    expect(dispatch).toHaveBeenCalledWith(sub);
  });

  it("dispatches function", () => {
    const dispatch = jest.fn();
    const func = () => {};
    const update = getUpdater(dispatch, () => func);
    expect(update("msg", "model")).toBe("model");
    expect(dispatch.mock.calls[0][0].effect).toBe(func);
  });

  it("dispatches promise", () => {
    const dispatch = jest.fn();
    const promise = Promise.resolve("result");
    const update = getUpdater(dispatch, () => promise);
    expect(update("msg", "model")).toBe("model");
    expect(dispatch.mock.calls[0][0].effect).toBe(promise);
  });

  it("handles update with effects", () => {
    const dispatch = jest.fn();
    const cmd = Cmd("effect");
    const sub = Sub.interval("effect", 100);
    const func = () => {};
    const promise = Promise.resolve("result");
    const obj = {};
    const update = getUpdater(dispatch, (msg, model) =>
      Update({ msg, model }, cmd, sub, func, promise, obj)
    );
    expect(update("msg", "model")).toMatchSnapshot();

    const { calls } = dispatch.mock;
    expect(calls[0][0]).toBe(cmd);
    expect(calls[1][0]).toBe(sub);
    expect(calls[2][0].effect).toBe(func);
    expect(calls[3][0].effect).toBe(promise);
    expect(calls[4][0].effect).toBe(obj);
  });

  it("throws error with invalid update", () => {
    expect(getUpdater(jest.fn(), "yolo")).toThrowError();
  });
});
