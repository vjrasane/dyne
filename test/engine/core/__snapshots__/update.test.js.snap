// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update handles basic update 1`] = `
Object {
  "model": "model",
  "msg": "msg",
}
`;

exports[`update handles update with effects 1`] = `
Object {
  "model": "model",
  "msg": "msg",
}
`;

exports[`update wraps a command 1`] = `
Return {
  "bind": [Function],
  "effects": Array [
    Command {
      "effect": "effect",
      "execute": [Function],
      "failure": [Function],
      "success": [Function],
    },
  ],
  "map": [Function],
  "model": null,
}
`;

exports[`update wraps a model 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": "model",
}
`;

exports[`update wraps a subscription 1`] = `
Return {
  "bind": [Function],
  "effects": Array [
    Subscription {
      "cancel": [Function],
      "canceller": [Function],
      "effect": [Function],
      "execute": [Function],
    },
  ],
  "map": [Function],
  "model": null,
}
`;
